### ISSUE TITLE
[2.2.21]"RuntimeError: maximum recursion depth exceeded" if events > 999

### ISSUE THREAD

**Issue by yuxisun1217** [Azure/WALinuxAgent#1035]
Description of problem:
If events > 999, WALA keeps printing "WARNING Too many files under: /var/lib/waagent/events, removing oldest" and finally raise exception: "RuntimeError: maximum recursion depth exceeded"
WALA-2.2.20 doesn't have this issue

Version-Release number of selected component (if applicable):
WALinuxAgent-2.2.21

RHEL Version:
RHEL-7.4/6.9
How reproducible: 
100%

Steps to Reproduce: 
1. Prepare a RHEL7.4 VM in Azure. Install WALA-2.2.21
2. Create 1000 files under /var/lib/waagent/events:
 service waagent stop
 cd /var/lib/waagent/events
 touch test{0001..1000}
 service waagent start
3. Check /var/log/waagent

Actual results: 
```
2018/02/07 21:54:56.289956 INFO Start env monitor service.
2018/02/07 21:54:56.342659 WARNING Too many files under: /var/lib/waagent/events, removing oldest
2018/02/07 21:54:56.438029 WARNING Too many files under: /var/lib/waagent/events, removing oldest
...(hundreds of the same log)
2018/02/07 21:55:19.748939 WARNING Too many files under: /var/lib/waagent/events, removing oldest
2018/02/07 21:55:19.820634 ERROR Failed to run 'run-exthandlers': Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/agent.py", line 167, in main
    agent.run_exthandlers()
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/agent.py", line 132, in run_exthandlers
    update_handler.run()
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/ga/update.py", line 308, in run
    logger.warn(msg)
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/common/logger.py", line 195, in warn
    DEFAULT_LOGGER.warn(msg_format, *args)
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/common/logger.py", line 62, in warn
    self.log(LogLevel.WARNING, msg_format, *args)
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/common/logger.py", line 87, in log
    appender.write(level, log_item)
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/common/logger.py", line 145, in write
    self.event_func(level, msg)
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/common/event.py", line 301, in add_log_event
    reporter.add_log_event(level, message)
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/common/event.py", line 248, in add_log_event
    self.save_event(json.dumps(data))
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/common/event.py", line 158, in save_event
    logger.warn("Too many files under: {0}, removing oldest".format(self.event_dir))
......
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/common/event.py", line 239, in add_log_event
    event = TelemetryEvent(7, "FFF0196F-EE4C-4EAF-9AA5-776F622DEB4F")
  File "/usr/lib/python2.6/site-packages/azurelinuxagent/common/protocol/restapi.py", line 288, in __init__
    self.parameters = DataContractList(TelemetryEventParam)
RuntimeError: maximum recursion depth exceeded
```

Expected results:
Can remove the oldest event.

**Comment by boumenot** [comment]
@yuxisun1217 - thank you for bringing this to my attention.  I thought we had an intergration test for this, but clearly not.  I will fix this in the next release.

**Comment by boumenot** [comment]
We introduced a new logger for the level WARNING and ERROR.  It looks like we get into a situation were the various logging statements just keep calling each other, and then cause us to blow up.

**Comment by boumenot** [comment]
I was able to reproduce the issue. At first, I didn't follow the instructions exactly, and it worked.  I tried it on a few different flavors of Linux, and they all worked too. I then realized I failed to stop the agent then create the 1,000+ files.

This greatly reduces the risk because we process new telemetry every few seconds, so the period of time when we could get stuck in this state is rather small.  Regardless, we've hotfixed the issue, and we are now rolling out 2.2.22.

**Comment by boumenot** [comment]
I was able to reproduce the issue. At first, I didn't follow the instructions exactly, and the agent worked correctly.  I tried it on a few different flavors of Linux, and they all worked too. I then realized I failed to stop the agent then create the 1,000+ files.

This greatly reduces the risk because we process new telemetry every few seconds, so the period of time when we could get stuck in this state is rather small.  Regardless, we've hot fixed the issue, and we are now rolling out 2.2.22.
